### ISSUE TITLE
Another test script for numerical stability

### ISSUE THREAD

**Issue by acgtyrant** [vacancy/Synchronized-BatchNorm-PyTorch#14]
Create a virtual environment and install torch and pytest, run pytest immediately to test the script below:

```
import torch
from torch import nn
from torch.nn import init

import pytest


def allclose(x, y):
    adiff = (x - y).abs().max()
    if (y == 0).all():
        rdiff = 'NaN'
    else:
        rdiff = (adiff / y).abs().max()
    message = (
            'Tensor close check failed\n'
            'adiff={}\n'
            'rdiff={}\n'
    ).format(adiff, rdiff)
    assert torch.allclose(x, y), message


class TBatchNorm(nn.Module):

    def __init__(
            self,
            num_features,
            eps=1e-5,
            momentum=0.1):
        super(TBatchNorm, self).__init__()
        self.num_features = num_features
        self.eps = eps
        self.momentum = momentum
        self.weight = nn.Parameter(torch.Tensor(num_features))
        self.bias = nn.Parameter(torch.Tensor(num_features))
        self.register_buffer('running_mean', torch.zeros(num_features))
        self.register_buffer('running_var', torch.ones(num_features))
        self.reset_parameters()

    def reset_running_stats(self):
        self.running_mean.zero_()
        self.running_var.fill_(1)

    def reset_parameters(self):
        self.reset_running_stats()
        nn.init.uniform_(self.weight)
        nn.init.zeros_(self.bias)

    def forward(self, input_):
        batchsize, channels, height, width = input_.size()
        numel = batchsize * height * width
        input_ = input_.permute(1, 0, 2, 3).contiguous().view(channels, numel)
        sum_ = input_.sum(1)
        sum_of_square = input_.pow(2).sum(1)
        mean = sum_ / numel
        sumvar = sum_of_square - sum_ * mean
        unbias_var = sumvar / (numel - 1)
        bias_var = sumvar / numel
        std = bias_var.clamp(self.eps) ** 0.5

        self.running_mean = (
                (1 - self.momentum) * self.running_mean
                + self.momentum * mean.detach())
        self.running_var = (
                (1 - self.momentum) * self.running_var
                + self.momentum * unbias_var.detach())

        output = (
                (input_ - mean.unsqueeze(1)) / std.unsqueeze(1) *
                self.weight.unsqueeze(1) + self.bias.unsqueeze(1))

        return output.permute(1, 0).contiguous().view(
                batchsize, channels, height, width)


@pytest.fixture(scope='module')
def instance():
    CHANNELS = 16
    batchnorm1 = nn.BatchNorm2d(CHANNELS, momentum=1)
    optimizer1 = torch.optim.SGD(batchnorm1.parameters(), lr=0.01)
    batchnorm2 = TBatchNorm(CHANNELS, momentum=1)
    with torch.no_grad():
        batchnorm2.weight = nn.Parameter(batchnorm1.weight.clone())
        batchnorm2.bias = nn.Parameter(batchnorm1.bias.clone())
    optimizer2 = torch.optim.SGD(batchnorm2.parameters(), lr=0.01)

    for _ in range(100):
        input_ = torch.rand(16, CHANNELS, 16, 16)

        input1 = input_.clone().requires_grad_(True)
        output1 = batchnorm1(input1)
        output1.sum().backward()
        optimizer1.step()

        input2 = input_.clone().requires_grad_(True)
        output2 = batchnorm2(input2)
        output2.sum().backward()
        optimizer2.step()

    return input1, batchnorm1, output1, input2, batchnorm2, output2


def test_input(instance):
    input1, batchnorm1, output1, input2, batchnorm2, output2 = instance
    allclose(input1, input2)


def test_output(instance):
    input1, batchnorm1, output1, input2, batchnorm2, output2 = instance
    allclose(output1, output2)


def test_input_grad(instance):
    input1, batchnorm1, output1, input2, batchnorm2, output2 = instance
    allclose(input1.grad, input2.grad)


def test_weight_grad(instance):
    input1, batchnorm1, output1, input2, batchnorm2, output2 = instance
    allclose(batchnorm1.weight.grad, batchnorm2.weight.grad)


def test_bias_grad(instance):
    input1, batchnorm1, output1, input2, batchnorm2, output2 = instance
    allclose(batchnorm1.bias.grad, batchnorm2.bias.grad)
```

Is this script reasonable?

**Comment by vacancy** [comment]
Thanks for the contribution! So, is this for comparing the numeric stability between `torch.nn.BatchNorm` and a self-implemented BatchNorm version?

**Comment by acgtyrant** [comment]
This is comparing the numeric stability between torch.nn.BatchNorm and your implemented BatchNorm version.

**Comment by vacancy** [comment]
Looks nice! Can you help with creating a PR for this? Thanks!

**Comment by vacancy** [comment]
@acgtyrant I have created a new branch named `new-numeric-test`. Can you help me confirm the implementation of `tests/test_numeric_batchnorm_v2.py`? I can't get the desired output.

**Comment by acgtyrant** [comment]
OK, it will takes me some time to check it.

**Comment by acgtyrant** [comment]
I do not reproduce too. First the view of output is wrong, it should be fixed as `return output.view(channels, batchsize, height, width).permute(1, 0, 2, 3).contiguous()`. I find that if height and width of input is greater than 1 at the same time, or the size of the input is great, or the iteration is great, then the test maybe failed.

**Comment by acgtyrant** [comment]
After fix the wrong view of the output, I try to use my distributed bn to train drn now, if it work, then although it is not very numeric stability but it can work.
